samples/grpc/local-drone-control-scala/src/main/scala/drones/Drone.scala

```diff
@@ -0,0 +1,104 @@
+/*
+ * Copyright (C) 2009-2023 Lightbend Inc. <https://www.lightbend.com>
+ */
+package drones
+
+import akka.Done
+import akka.actor.typed.ActorRef
+import akka.actor.typed.ActorSystem
+import akka.actor.typed.Behavior
+import akka.cluster.sharding.typed.scaladsl.ClusterSharding
+import akka.cluster.sharding.typed.scaladsl.Entity
+import akka.cluster.sharding.typed.scaladsl.EntityTypeKey
+import akka.persistence.typed.PersistenceId
+import akka.persistence.typed.scaladsl.Effect
+import akka.persistence.typed.scaladsl.EventSourcedBehavior
+import akka.persistence.typed.scaladsl.ReplyEffect
+
+import scala.collection.immutable.Queue
+
+object Drone {
+  sealed trait Command
+
+  final case class ReportPosition(position: Position, replyTo: ActorRef[Done])
+      extends Command
+
+  sealed trait Event
+  final case class PositionUpdated(position: Position) extends Event
+  final case class CoarseGrainedLocationChanged(
+      coordinates: CoarseGrainedCoordinates)
+      extends Event
+
+  final case class State(
+      currentPosition: Option[Position],
+      historicalPositions: Queue[Position]) {
+    def coarseGrainedCoordinates: Option[CoarseGrainedCoordinates] =
+      currentPosition.map(p =>
+        CoarseGrainedCoordinates.fromCoordinates(p.coordinates))
+  }
+  private val emptyState = State(None, Queue.empty)
+
+  val EntityKey: EntityTypeKey[Command] =
+    EntityTypeKey[Command]("Drone")
+
+  val LocationHistoryLimit = 100
+
+  def init(system: ActorSystem[_]): Unit = {
+    ClusterSharding(system).init(Entity(EntityKey)(entityContext =>
+      Drone(entityContext.entityId)))
+  }
+
+  def apply(entityId: String): Behavior[Command] =
+    EventSourcedBehavior[Command, Event, State](
+      PersistenceId(EntityKey.name, entityId),
+      emptyState,
+      handleCommand,
+      handleEvent)
+
+  private def handleCommand(
+      state: State,
+      command: Command): ReplyEffect[Event, State] = command match {
+    case ReportPosition(position, replyTo) =>
+      if (state.currentPosition.contains(position))
+        // already seen
+        Effect.none.thenReply(replyTo)(_ => Done)
+      else {
+        val newCoarseGrainedLocation =
+          CoarseGrainedCoordinates.fromCoordinates(position.coordinates)
+        if (state.coarseGrainedCoordinates.contains(newCoarseGrainedLocation)) {
+          // same grid location as before
+          Effect
+            .persist(PositionUpdated(position))
+            .thenReply(replyTo)(_ => Done)
+        } else {
+          // new grid location
+          Effect
+            .persist(
+              PositionUpdated(position),
+              CoarseGrainedLocationChanged(newCoarseGrainedLocation))
```

Comment:
later, to illustrate starting from snapshots we could snapshot when there is a CoarseGrainedLocationChanged